design-document.md

```diff
@@ -232,6 +232,352 @@ Note: Two non-colluding aggregators --- one leader and one standard aggregator
+
+Next, the client encodes its measurements as an input to the given PA protocol.
+It then generates a validity proof for its input and uses `context` to split the
+input and proof into a leader share and a helper share, where the latter is
+encrypted by the HPKE context. (Note that the secret-sharing and encryption
+procedures are specific to the PA protocol.)
+
+[[OPEN ISSUE: Is it safe to generate the proof once, then secret-share between
+each (leader, helper) pair? Probably not in general, but maybe for Prio?]]
+
+Finally, the clients sends an HTTP request to `[leader]/upload` with the
+following payload:
+
+```
+struct {
+  PATask task;                // PAParameter.task
+  uint8 helper_key_config_id; // HPKEKeyConfig.id
+  Url helper_url;             // URL of helper endpoint
+  PAHelperShare helper_share; // The helper's encrypted share
+  PALeaderShare leader_share; // The leader's share
+} PAUplaod;
+
+struct {
+  opaque enc<0..2^16-1>;     // Encapsulated HPKE context
+  opaque payload<0..2^24-1>; // The encrypted share
+} PAHelperShare;
+
+opaque PALeaderShare<0..2^24-1>;
+```
+
+### Input validation
+
+The leader handles HTTP requests to `[leader]/upload` as follows. It first
+checks the request for a well-formed `PAUpload` message. If not found, then the
+leader aborts and alerts the client with "malformed upload". Otherwise, if a
+well-formed `PAUpload` message is found, the leader responds to the request with
+"200 OK".
+
+#### Verify Start
+
+The leader begins the input-validation protocol by looking up the PA parameters
+`PAParam` for which `PAUpload.task == PAParam.task` and `PAUpload.helper_url` is
+in `PAParam.helper_urls`. If none are found, then the leader aborts without
+alerting the client.
+
+**Request.**
+Let `[helper]` denote the URL corresponding to `PAUpload.helper_url`. The leader
+sends an HTTP request to `[helper]/verify_start` with the following payload:
+
+```
+struct {
+  PATask task;                           // PAUpload.task
+  uint8 key_config_id;                   // PAUpload.helper_key_config_id
+  PAHelperShare shares<0..2^24-1>;       // Any number of helper shares
+  select (PAVerifyStartReq.task.proto) { // Payload for each helper share
+    case prio: PrioVerifyStartReq payloads<0..2^16-1>;
+    case hits: HitsVerifyStartReq payloads<0..2^16-1>;
+  }
+} PAVerifyStartReq;
+```
+
+This message contains any number of encrypted helper shares: it might contain
+the shares uploaded by a single client, or it might contain the shares uploaded
+by multiple clients; we only require that all of the shares correspond to the
+same PA task and HPKE key configuration. This allows the aggregator to
+batch-process a number of inputs at once.
+
+Each helper share has a corresponding message payload contained in
+`PAVerifyStartReq.shares`. This includes any protocol-specific information the
+helper needs for input validation, e.g., the joint randomness used by the leader
+and helper for the protocol run.
```

Comment:
How can `PAVerifyStartReq.shares` carry protocol-specific information if it only includes `enc` and `payload`? Is `payload` the share encryption and then other goop? (Reading on, that does seem to be the case. I'm not sure how we'd  clarify that now.)

Reply:
Yes that's the case. `PAVerifyStartREq.shares` should be `PAVerifyStartReq.payloads` here.